diff/highlight: diminish low-priority CursorLine

justinmk opened the pull request:
Render low-priority CursorLine only as underline when the cursor is on a diff line. This is because typically a diff line has a foreground highlight, which makes text difficult to see if CursorLine has a background highlight.

CursorLine highlight with ctermfg/guifg=NONE is "low priority" in the presence of 'colorcolumn' and some other highlights. But before this PR it was not altered in a diff line. 

To see the effect, change `ctermfg=white` to `ctermfg=NONE` in this command:

    nvim -u NORC +'hi CursorLine ctermbg=darkgrey ctermfg=white' +'set cursorline' +'call feedkeys("ifoo\<CR>foo\<CR>foo\<CR>bar", "x")'  +vnew +'call feedkeys("ibee\<CR>foo\<CR>foo\<CR>foo\<CR>baz", "x")' +'windo diffthis'

ref #6380

todo:

- [x] tests

cc @zhou13

justinmk committed:
diff/highlight: do not overlay low-priority CursorLine

ref #6380

zhou13 commented:
Does that mean once `CursorLine` is set, other line-wised highlight will be ignored?

justinmk commented:
Can you give an example?

zhou13 commented:
For example, if `cursorline` is set, will nvim ignore the foreground of VimDiff?

justinmk committed:
diff/highlight: Show underline for low-priority CursorLine

justinmk committed:
refactor: Rename get_term_attr_entry

Rename get_term_attr_entry to hl_get_term_attr, similar to
hl_get_syn_attr, hl_get_ui_attr.

justinmk commented on src/nvim/highlight.c:
```diff
@@ -177,8 +177,22 @@ void update_window_hl(win_T *wp, bool invalid)
+int hl_get_underline(void)
+{
+  int hl_attrs = HL_UNDERLINE;
+  return hl_get_term_attr(&(HlAttrs) {
```
using `kHlTerminal` here. Not sure if it matters.

bfredl replied:
should be `kHlUI` if the trick is used, but I think it can be avoided entirely (see other comment).

bfredl commented on src/nvim/screen.c:
```diff
@@ -3052,9 +3052,11 @@ win_line (
+        // Overlay CursorLine onto diff-mode highlight.
+        if (wp->w_p_cul && lnum == wp->w_cursor.lnum) {
+          line_attr = 0 != line_attr_lowprio  // Low-priority CursorLine
+            ? hl_combine_attr(line_attr, hl_get_underline())
```
Why not `hl_combine_attr(win_hl_attr(wp, HLF_CUL), line_attr)` which would work for any attribute and not only `underline` ?

justinmk replied:
@bfredl Thanks. That effectively hides the CursorLine with the use-case given above, but maybe we could do that + add underline. The idea is to have something that's visible without being unreadable.

bfredl replied:
@justinmk How could `hl_combine_attr(win_hl_attr(wp, HLF_CUL), line_attr)`  "effectively hide underline" in any way that `hl_combine_attr(line_attr, hl_get_underline())` doesn't. The former will include _any_ attribute that CursorLine contains.

bfredl replied:
If the purpose rather is to _actively_ special case CursorLine on diff lines, then `CursorLineDiff` group would be the proper and configurable option (which even can default to underline, for this behavior by default) , rather than adding _ad-hoc_ logic for hardcoding an implicit `gui=underline` group, which is not done anywhere else. What is the motivation to add code to _prohibit_ users to configure this particular case?

justinmk replied:
I mean visually, the diff highlight still dominates. Try the command given in the top description.

bfredl replied:
Again, that depends on the attributes used. If `reverse` is used, both high and low priority already works, but with this change diff lines are needlessly different.

justinmk replied:
Yes, it depends on the attributes.

The point of this PR is to avoid a degenerate case for something like `hi CursorLine ctermbg=grey`. With `hi DiffAdd ctermfg=black ctermbg=red`, that's hard to read.

bfredl replied:
Yes, all I'm saying is if the corner case is a result of user configuration,  the override itself should be configurable (compare CursorLineNr). Otherwise it will just make some configurations better by making others worse..

bfredl replied:
> But with the latest push AFAIK it won't make anything worse except if someone considers it worse to have an underline. It's also overridable by setting foreground on CursorLine (so it's not "low priority").

@justinmk All other screen attributes added by vim/nvim are configurable by the user, moving towards a mixture of user configurable attributes and hard-coded attributes is a step backwards, and an unexpected one.  If it counts for anything, it also breaks my setup (i don't use `CursorLine` always, but when I do), by not only unexpected underline but also unreadable text on diff lines. And yes it is "fixable" by setting guifg, but why should I or any user need to use guifg if we only want guibg (or some non-ul attribute)? Why should anyone be forced to use guifg to take control of bg and attributes, instead of having the later being controlled by `guibg=` and `gui=` like in all other UI highlights in nvim?

justinmk committed:
preserve_exit: Ignore SIGHUP

closes #9274
ref #9028

If stdin closed then read_error_exit calls preserve_exit. Handling
SIGHUP during preserve_exit would cause a premature teardown, and
conflicts with e.g. ui_bridge_stop which waits for TUI to teardown.

Vim ignores SIGHUP in its prepare_to_exit and getout_preserve_modified
routines:

    /* Ignore SIGHUP, because a dropped connection causes a read error, which
     * makes Vim exit and then handling SIGHUP causes various reentrance
     * problems. */
    signal(SIGHUP, SIG_IGN);
